src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -41,10 +45,94 @@ public RouteEndpointDataSource(IServiceProvider applicationServices, bool throwO
             HttpMethods = httpMethods,
             RouteAttributes = RouteAttributes.None,
             Conventions = conventions,
-            FinallyConventions = finallyConventions
+            FinallyConventions = finallyConventions,
+            InferMetadataFunc = null, // Metadata isn't infered from RequestDelegate endpoints
+            CreateHandlerRequestDelegateFunc = CreateHandlerRequestDelegate
         });
 
         return new RouteHandlerBuilder(conventions, finallyConventions);
+
+        static RequestDelegate CreateHandlerRequestDelegate(Delegate handler, RequestDelegateFactoryOptions options, RequestDelegateMetadataResult? metadataResult)
+        {
+            var requestDelegate = (RequestDelegate)handler;
+
+            // Use request delegate directly when there are no filters.
+            if (options.EndpointBuilder is null || options.EndpointBuilder.FilterFactories.Count == 0)
+            {
+                return requestDelegate;
+            }
+
+            return BuildFilterPipeline(requestDelegate, options);
+        }
+    }
+
+    private static RequestDelegate BuildFilterPipeline(RequestDelegate requestDelegate, RequestDelegateFactoryOptions options)
+    {
+        Debug.Assert(options.EndpointBuilder != null);
+
+        var serviceProvider = options.ServiceProvider ?? options.EndpointBuilder.ApplicationServices;
+        var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? new JsonOptions();
```

Comment:
```suggestion
        var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? JsonOptionsDefaultSerializerOptions;
```

?? Maybe wait for @brunolins16 's change here?

Reply:
That doesn’t compile. It’s internal in another project

Reply:
It’s why https://github.com/dotnet/aspnetcore/issues/46024 exists. A lot of this becomes much simpler

Reply:
I see